### PR TITLE
Refactoring; splitting up modelDescription & terminalsAndIcons parsing

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**
Refactoring:
1. Removed common (xml scheme) attributes list, instead duplicated for terminalsAndIcons (for simplicity)
2. Split all scheme related enums, global maps into modelDescription, terminalsAndIcons
3. Replaced element/attr enums by appropriate unions. Introduced new macros for easy allocation: FMI3_ATTR_ANY for xml independent contexts, FMI_ATTR_TERMICON for terminalsAndIcons, FMI3_ATTR for modelDescription (FMI3_ATTR_MODELDESCRIPTION would artificially blow up lines and the 3 implicitly points to FMI3 modelDescription)
4. Added fmi3_xml_type enum for context struct
5. All accesses to global structs are now via functions and use appropriate switch (xmlType) case ...

Next step: 
- Re-location of code into more suitable logical file structures
- Clean-up of imports

### CODE REVIEW

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 1]
```diff
@@ -41,45 +42,143 @@ typedef unsigned long long int  fmi3_uint_buf_t; /* same size as fmi3_int_buf_t
...
+fmi_xml_scheme_termIcon_info_t fmi_xml_scheme_termIcon_info[fmi_xml_elm_termIcon_number] = {
+    FMI_XML_ELMLIST_TERM_ICON(EXPAND_ELM_SCHEME_TERMICON)
+    {fmi_xml_elm_termIcon_actual_number, 0, 0},
+    FMI_XML_ELMLIST_ALT_TERM_ICON(EXPAND_ELM_SCHEME_TERMICON)
+};
+
+// TODO: Move to more suitable place?
+static fmi3_xml_scheme_info_t fmi3_xml_get_scheme_info(fmi3_xml_parser_context_t* context, fmi3_xml_elm_t enu) {
+    const fmi3_xml_type_t xmlType = context->xmlType;
+
```
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_parser.c:90]: Contents of this function likely a bit more "verbose" than strictly required, but I don't necessarily think this is an issue; makes the conversion very explicit.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 2]
```diff
@@ -923,37 +1030,70 @@ jm_vector(char)* fmi3_xml_reserve_parse_buffer(fmi3_xml_parser_context_t* contex
...
     if (!context->attrMapById) return -1;
-    context->attrMapByName = jm_vector_alloc(jm_named_ptr)(fmi3_xml_attr_number, fmi3_xml_attr_number, context->callbacks);
+    context->attrMapByName = jm_vector_alloc(jm_named_ptr)(mapSize, mapSize, context->callbacks);
     if (!context->attrMapByName) return -1;
-    for(i = 0; i < fmi3_xml_attr_number; i++) {
+    for(i = 0; i < mapSize; i++) {
         jm_named_ptr map;
         jm_vector_set_item(jm_string)(context->attrMapById, i, 0);
-        map.name = fmi3_xmlAttrNames[i];
+        map.name = fmi3_xml_get_xml_attr_name(context, FMI3_ATTR_ANY(i));
```
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_parser.c:1072]: Example for generic parser context, hence using FMI3_ATTR_ANY here.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 3]
```diff
@@ -1290,6 +1430,24 @@ static void XMLCALL fmi3_parse_element_data(void* c, const XML_Char *s, int len)
         }
 }
 
+/* Allocate fmi3_xml_parser_context_t with a given xmlType */
+static fmi3_xml_parser_context_t* fmi3_xml_allocate_parser_context(
```
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_parser.c:1434]: I don't think there is any other way to do this really. It is a bit of a hack, but I do see value in actually having the xmlType as const in the structure, we want to make sure it does not change.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 4]
```diff
@@ -29,6 +29,7 @@
 #include "../FMI/fmi_xml_terminals_and_icons_impl.h"
 #include "fmi3_xml_parser.h"
 #include "JM/jm_portability.h"
+#include "fmi3_xml_parser_context_impl.h" // TODO: Really needed here?
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_parser.c:32]: Is the comment still relevant?

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 5]
```diff
@@ -923,37 +1030,70 @@ jm_vector(char)* fmi3_xml_reserve_parse_buffer(fmi3_xml_parser_context_t* contex
...
     int i;
-    context->attrMapById = jm_vector_alloc(jm_string)(fmi3_xml_attr_number, fmi3_xml_attr_number, context->callbacks);
+    size_t mapSize = fmi3_get_enum_size(context);
+    context->attrMapById = jm_vector_alloc(jm_string)(mapSize, mapSize, context->callbacks);
     if (!context->attrMapById) return -1;
-    context->attrMapByName = jm_vector_alloc(jm_named_ptr)(fmi3_xml_attr_number, fmi3_xml_attr_number, context->callbacks);
+    context->attrMapByName = jm_vector_alloc(jm_named_ptr)(mapSize, mapSize, context->callbacks);
     if (!context->attrMapByName) return -1;
-    for(i = 0; i < fmi3_xml_attr_number; i++) {
+    for(i = 0; i < mapSize; i++) {
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_parser.c:1069]:
```suggestion
    for (i = 0; i < mapSize; i++) {
```

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 6]
```diff
@@ -923,37 +1030,70 @@ jm_vector(char)* fmi3_xml_reserve_parse_buffer(fmi3_xml_parser_context_t* contex
...
+// TODO: Move to better place?
+static int fmi3_create_elm_map(fmi3_xml_parser_context_t* context) {
+    int i;
+    size_t mapSize = fmi3_get_enum_size(context);
+    size_t mapSizeActual = fmi3_get_enum_size_actual(context);
+    context->elmMap = jm_vector_alloc(fmi3_xml_element_handle_map_t)(mapSizeActual, mapSize, context->callbacks);
     if (!context->elmMap) return -1;
-    for(i = 0; i < fmi3_xml_elm_actual_number; i++) {
-        fmi3_xml_element_handle_map_t item = fmi3_element_handle_map[i];
+    for(i = 0; i < mapSizeActual; i++) {
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_parser.c:1087]:
```suggestion
    for (i = 0; i < mapSizeActual; i++) {
```

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 7]
```diff
@@ -1298,13 +1456,11 @@ int fmi3_xml_parse_model_description(fmi3_xml_model_description_t* md,
...
-    context = (fmi3_xml_parser_context_t*)md->callbacks->calloc(1, sizeof(fmi3_xml_parser_context_t));
-    if (!context) {
-        jm_log_fatal(md->callbacks, "FMIXML", "Could not allocate memory for XML parser context");
-    }
-    context->callbacks = md->callbacks;
+    context = fmi3_xml_allocate_parser_context(md->callbacks, fmi3_xml_type_modelDescription);
+
     context->modelDescription = md;
     if (fmi3_xml_alloc_parse_buffer(context, 16)) return -1;
+    // if (fmi3_create_modelDescription_attr_map(context) || fmi3_create_modelDescription_elm_map(context)) {
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_parser.c:1463]: Should we remove this code/comment now?

File: src/XML/src/FMI3/fmi3_xml_parser_context_impl.h [hunk 8]
```diff
@@ -28,23 +28,50 @@
...
 
+
 jm_vector_declare_template(fmi3_xml_element_handle_map_t)
 
 /**
  * Struct for saving and accessing data between element handlers.
  */
 struct fmi3_xml_parser_context_t {
 
+    /* Flag for XML being handled, MUST BE FIRST */
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_parser_context_impl.h:72]: I think we could add in this comment that it is either `...` or `...` depending on what we are parsing.

**Review by modelonrobinandersson — APPROVED** [review]
I think this all looks very good, well done!
I only noticed some very minor things (feel free to ignore).

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 4 above)
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c:32]: Yes, somewhat, will be dealt with in follow-up cleanup.

File: src/XML/src/FMI3/fmi3_xml_parser_context_impl.h (hunk 8 above)
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser_context_impl.h:72]: Expanded the comment

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 7 above)
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c:1463]: Good catch, removed.

**Comment by PeterMeisrimelModelon** [comment]
@modelonrobinandersson See latest commit for a few more things renamed for clarity/consistency.

**Comment by modelonrobinandersson** [comment]
> @modelonrobinandersson See latest commit for a few more things renamed for clarity/consistency.

Looks great! I think the renaming also helped a lot.

**Pull request merged**